simperby/src/node/stf.rs

```diff
@@ -0,0 +1,288 @@
+///
+/// Note that it does not interact with the storage checkpoint;
+/// the caller must check the result and
+/// revert the state if it is not successful.
+pub(super) async fn execute_block(
+    storage: &mut StateStorage,
+    last_header: BlockHeader,
+    transactions: Vec<Transaction>,
+    header_info: HeaderInfo,
+) -> Result<BlockHeader, ExecutionError> {
+    let mut next_version = last_header.version.clone();
+    for (index, transaction) in transactions.iter().enumerate() {
+        if let Some(state_transition) = &transaction.state_transition {
+            match state_transition {
+                StateTransition::AddValidator {
+                    public_key,
+                    voting_power,
+                } => {
+                    execute_add_validator(storage, index, public_key.clone(), *voting_power)
+                        .await?;
+                }
+                StateTransition::RemoveValidator { public_key } => {
+                    execute_remove_validator(storage, index, public_key.clone()).await?;
+                }
+                StateTransition::Delegate {
+                    delegator,
+                    delegatee,
+                    target_height,
+                    commitment,
+                } => {
+                    execute_delegate(
+                        storage,
+                        &last_header,
+                        index,
+                        delegator.clone(),
+                        delegatee.clone(),
+                        *target_height,
+                        commitment.clone(),
+                    )
+                    .await?;
+                }
+                StateTransition::Undelegate { delegator } => {
+                    execute_undelegate(storage, index, delegator.clone()).await?;
+                }
+                StateTransition::InsertData { key, value } => {
+                    storage.insert_record(Hash256::hash(key), value).await?;
+                }
+                StateTransition::RemoveData(key) => {
+                    storage.remove_record(Hash256::hash(key)).await?;
+                }
+                StateTransition::UpdateVersion { version } => {
+                    let ver1 = semver::Version::parse(&last_header.version).map_err(|_| {
+                        ExecutionError::InvalidHeader(format!(
+                            "invalid version: {}",
+                            last_header.version
+                        ))
+                    })?;
+                    let ver2 = semver::Version::parse(version).map_err(|_| {
+                        ExecutionError::InvalidHeader(format!(
+                            "invalid version: {}",
+                            last_header.version
+                        ))
+                    })?;
+                    if ver2 <= ver1 {
+                        return Err(ExecutionError::InvalidHeader(format!(
+                            "version must grow: {}",
```

Comment:
How about also logging `ver1` and showing both of the versions? I think it would be a better error message.

Reply:
fixed